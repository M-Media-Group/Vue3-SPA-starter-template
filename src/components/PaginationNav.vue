<script setup lang="ts">
import { computed } from "vue";
import BaseButton from "./BaseButton.vue";
import i18n from "@/locales/i18n";

const emit = defineEmits([
  /** The page the user has navigated to, either by clicking directly on a page or by using the previous and next buttons */
  "update:currentPage",
]);

const props = defineProps({
  /** The count of total items */
  totalItems: {
    type: Number,
    required: true,
    validator: (value: number) => value >= 0,
  },
  /** The current page. Note that this is the same as the emit, so when using this component, you can just use a v-model:currentPage for two-way binding */
  currentPage: {
    type: Number,
    required: false,
    default: 1,
  },
  /** The results to show per page */
  resultsPerPage: {
    type: Number,
    required: false,
    default: 5,
  },
  /** The maximum amount of pages to show before truncating. Basically, it will be: [min-page] ... [x amount of pages around and including the current page, determined by this prop] ... [max-page] */
  maxPages: {
    type: Number,
    required: false,
    default: 5,
    validator: (value: number) => value >= 3,
  },
  /** Show the previous and next buttons */
  showPrevNext: {
    type: Boolean,
    required: false,
    default: true,
  },
  /** An optional lastPage parameter to override the calculated last page. If this value is set and less than the calculated last page, then the pagination will show the last page as this value */
  lastPage: {
    type: Number,
    required: false,
  },
  /** The text to use in the separator */
  separatorText: {
    type: String,
    required: false,
    default: "...",
  },
  /** Go to page input */
  goToPage: {
    type: Boolean,
    required: false,
    default: false,
  },
});

const totalPages = computed(() => {
  const computedTotal = Math.ceil(props.totalItems / props.resultsPerPage);
  return props.lastPage
    ? Math.min(computedTotal, props.lastPage)
    : computedTotal;
});

type Page = {
  pageNumber: number;
  text: string;
  clickable: boolean;
  selected?: boolean;
};

const formatNumer = (number: number) => {
  return new Intl.NumberFormat(i18n.global.locale.value).format(number);
};

/** An array of pages to show. We show x-1, then a separator, then the last page,. The first part length is the maxPages - 1 */
const pages = computed(() => {
  const pages: Page[] = [];
  const maxPages = props.maxPages - 1;
  const currentPage = props.currentPage;
  const newTotalPages = totalPages.value;
  const halfMaxPages = Math.floor(maxPages / 2);

  // If the maxPages + 2 (the first and last page) is greater than the total pages, then we just show all the pages
  if (newTotalPages <= maxPages + 2) {
    for (let i = 1; i <= newTotalPages; i++) {
      pages.push({ pageNumber: i, text: formatNumer(i), clickable: true });
    }
    return pages;
  }

  const showFirstSeparator =
    currentPage > halfMaxPages &&
    // Or if we already will show the page in the second part, (e.g. to prevent 1 ... 1 2 3 4 5 ... 10)
    currentPage > halfMaxPages + 1 &&
    newTotalPages > maxPages;
  const showLastSeparator =
    currentPage < newTotalPages - halfMaxPages &&
    // Or if we already will show the page in the second part, (e.g. to prevent 1 ... 1 2 3 4 5 ... 10)
    currentPage < newTotalPages - halfMaxPages - 1 &&
    newTotalPages > maxPages;

  const firstPageToShow = showFirstSeparator ? currentPage - halfMaxPages : 1;
  const lastPageToShow = showLastSeparator
    ? currentPage + halfMaxPages
    : newTotalPages;
  if (showFirstSeparator) {
    pages.push({ pageNumber: 1, text: formatNumer(1), clickable: true });
    pages.push({ pageNumber: 0, text: props.separatorText, clickable: false });
  }
  for (let i = firstPageToShow; i <= lastPageToShow; i++) {
    pages.push({ pageNumber: i, text: formatNumer(i), clickable: true });
  }
  if (showLastSeparator) {
    pages.push({ pageNumber: 0, text: props.separatorText, clickable: false });
    pages.push({
      pageNumber: newTotalPages,
      text: formatNumer(newTotalPages),
      clickable: true,
    });
  }
  return pages;
});

const handleInput = (value: string) => {
  const number = parseInt(value, 10);
  if (number > 0 && number <= totalPages.value) {
    emit("update:currentPage", number);
  }
};
</script>
<template>
  <nav aria-label="Pagination">
    <template v-if="showPrevNext">
      <BaseButton
        @click="emit('update:currentPage', currentPage - 1)"
        aria-label="Previous page"
        class="prev"
        :disabled="currentPage <= 1"
      >
        <span aria-hidden="true">&laquo;</span>
      </BaseButton>
    </template>
    <template v-for="page in pages" :key="page.pageNumber">
      <BaseButton
        v-if="page.clickable"
        @click="emit('update:currentPage', page.pageNumber)"
        :aria-label="`Go to page ${page.pageNumber}`"
        :disabled="page.pageNumber === currentPage"
        class="outline contrast"
      >
        {{ page.text }}
      </BaseButton>
      <button v-else disabled class="outline secondary" type="button">
        {{ page.text }}
      </button>
    </template>
    <template v-if="goToPage">
      <input
        type="number"
        min="1"
        :max="totalPages"
        :value="currentPage"
        @input="handleInput(($event.target as HTMLInputElement).value)"
        style="width: 3rem"
        aria-label="Go to page"
      />
    </template>
    <template v-if="showPrevNext">
      <BaseButton
        @click="emit('update:currentPage', currentPage + 1)"
        aria-label="Next page"
        class="next"
        :disabled="currentPage >= totalPages"
      >
        <span aria-hidden="true">&raquo;</span>
      </BaseButton>
    </template>
  </nav>
</template>
