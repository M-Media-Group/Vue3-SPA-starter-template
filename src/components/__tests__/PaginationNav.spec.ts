import { describe, expect, it } from "vitest";

import { RouterLinkStub, mount } from "@vue/test-utils";
import PaginationNav from "../PaginationNav.vue";
import "html-validate/vitest";

describe("Pagination", () => {
  it("renders correctly by default", () => {
    const wrapper = mount(PaginationNav, {
      props: {
        totalItems: 100,
        resultsPerPage: 10,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    const pagination = wrapper.find("nav");
    expect(pagination.exists()).toBe(true);

    // The first button is the "prev" page and it should be disabled
    const prevButton = wrapper.find("button");
    expect(prevButton.attributes("disabled")).toBe("");
    // The last button is the "next" page and it should not be disabled
    const nextButton = wrapper.find("button:last-of-type");
    expect(nextButton.attributes("disabled")).toBe(undefined);
    // The second button is the first page and it should also be disabled since we are on it
    const firstButton = wrapper.find("button:nth-of-type(2)");
    expect(firstButton.attributes("disabled")).toBe("");
    // There should be 10 pages
    const buttons = wrapper.findAll("button");
    // With the default of maxPages: 5, we should have 7 buttons (5 pages + prev + next)
    expect(buttons.length).toBe(7);

    // @ts-ignore
    // @ts-ignore an ambient declaration file doesnt seem to work to add toHTMLValidate @todo and check
    expect(wrapper.html()).toHTMLValidate();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("does not show duplicate pages", () => {
    const wrapper = mount(PaginationNav, {
      props: {
        totalItems: 30,
        resultsPerPage: 10,
        currentPage: 3,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    // There should be 5 buttons
    const buttons = wrapper.findAll("button");
    expect(buttons.length).toBe(5);

    // None of the buttons should have the same text
    const texts = buttons.map((button) => button.text());
    expect(new Set(texts).size).toBe(texts.length);
  });

  it("does not show duplicate pages", () => {
    const wrapper = mount(PaginationNav, {
      props: {
        totalItems: 300,
        resultsPerPage: 10,
        currentPage: 3,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    // There should be 5 buttons
    const buttons = wrapper.findAll("button");

    // None of the buttons should have the same text
    const texts = buttons.map((button) => {
      // If its a "...", we skip it
      if (button.text() === "...") return;
      return button.text();
    });

    const textsSize = new Set(texts).size;
    expect(textsSize).toBe(texts.length);
  });

  it("does not show duplicate pages", () => {
    const wrapper = mount(PaginationNav, {
      props: {
        totalItems: 100,
        resultsPerPage: 10,
        currentPage: 10,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    // There should be 7 buttons
    const buttons = wrapper.findAll("button");
    expect(buttons.length).toBe(7);

    // None of the buttons should have the same text
    const texts = buttons.map((button) => {
      // If its a "...", we skip it
      if (button.text() === "...") return;
      return button.text();
    });

    const textsSize = new Set(texts).size;
    expect(textsSize).toBe(texts.length);
  });

  it("does not show useless separators towards the end", () => {
    const wrapper = mount(PaginationNav, {
      props: {
        totalItems: 40,
        resultsPerPage: 10,
        currentPage: 1,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    // There should be 7 buttons
    const buttons = wrapper.findAll("button");
    expect(buttons.length).toBe(6);

    // There should be no "..." in the buttons
    const texts = buttons.map((button) => button.text());
    expect(texts).not.toContain("...");
  });

  it("does not show useless separators towards the end", () => {
    const wrapper = mount(PaginationNav, {
      props: {
        totalItems: 50,
        resultsPerPage: 10,
        currentPage: 4,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    // There should be 7 buttons
    const buttons = wrapper.findAll("button");
    expect(buttons.length).toBe(7);

    // There should be no "..." in the buttons
    const texts = buttons.map((button) => button.text());
    expect(texts).not.toContain("...");
  });

  it("does not show useless separators towards the start", () => {
    const wrapper = mount(PaginationNav, {
      props: {
        totalItems: 40,
        resultsPerPage: 10,
        currentPage: 4,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    // There should be 7 buttons
    const buttons = wrapper.findAll("button");
    expect(buttons.length).toBe(6);

    // There should be no "..." in the buttons
    const texts = buttons.map((button) => button.text());
    expect(texts).not.toContain("...");
  });

  it("correctly renders really large page numbers", () => {
    const wrapper = mount(PaginationNav, {
      props: {
        totalItems: 1000000000,
        resultsPerPage: 10,
        currentPage: 500000,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    // There should be 7 buttons
    const buttons = wrapper.findAll("button");

    // The second to last button is the last page - it should read "100,000,000"
    const lastButton = buttons[buttons.length - 2];
    expect(lastButton.text()).toBe("100,000,000");
  });

  it("emits the correct page when a button is clicked", async () => {
    const wrapper = mount(PaginationNav, {
      props: {
        totalItems: 100,
        resultsPerPage: 10,
        currentPage: 1,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    const buttons = wrapper.findAll("button");
    // Go to page 3 button will have the text "3"
    const page3Button = buttons[3];
    await page3Button.trigger("click");
    expect(wrapper.emitted("update:currentPage")?.[0]).toEqual([3]);
  });

  it("emits the correct page when the next button is clicked", async () => {
    const wrapper = mount(PaginationNav, {
      props: {
        totalItems: 100,
        resultsPerPage: 10,
        currentPage: 1,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    const nextButton = wrapper.find("button:last-of-type");
    await nextButton.trigger("click");
    expect(wrapper.emitted("update:currentPage")?.[0]).toEqual([2]);
  });

  it("emits the correct page when the previous button is clicked", async () => {
    const wrapper = mount(PaginationNav, {
      props: {
        totalItems: 100,
        resultsPerPage: 10,
        currentPage: 2,
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    const prevButton = wrapper.find("button");
    await prevButton.trigger("click");
    expect(wrapper.emitted("update:currentPage")?.[0]).toEqual([1]);
  });
});
