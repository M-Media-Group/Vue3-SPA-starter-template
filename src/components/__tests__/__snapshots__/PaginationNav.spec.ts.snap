// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > renders correctly by default 1`] = `
"<nav aria-label="Pagination"><button disabled="" type="button" aria-label="Previous page" class="prev"><span aria-hidden="true">«</span></button><button disabled="" type="button" aria-label="Go to page 1" class="outline contrast">1</button><button type="button" aria-label="Go to page 2" class="outline contrast">2</button><button type="button" aria-label="Go to page 3" class="outline contrast">3</button><button disabled="" class="outline secondary" type="button">...</button><button type="button" aria-label="Go to page 10" class="outline contrast">10</button>
  <!--v-if--><button type="button" aria-label="Next page" class="next"><span aria-hidden="true">»</span></button>
</nav>"
`;
